v3/testdata/index_scan

```diff
@@ -0,0 +1,63 @@
+exec
+CREATE TABLE a (x INT PRIMARY KEY, y INT UNIQUE)
+----
+table a
+  x NOT NULL
+  y NULL
+  (x) KEY
+  (y) WEAK KEY
+
+exec
+CREATE TABLE b (x INT, z INT)
+----
+table b
+  x NULL
+  z NULL
+
+prep,memo,search
+SELECT * FROM a
+----
+1: [scan a] [index scan a@primary]
+
+prep,memo,search
+SELECT * FROM a, b
+----
+3: [inner join [1 2 -]] [inner join [2 1 -]]
+2: [scan b] [index scan b@primary]
+1: [scan a] [index scan a@primary]
+
+prep,memo,search
+SELECT * FROM a WHERE x > 1
+----
+5: [select [1 4]]
+4: [comp (>) [2 3]]
+3: [const 1]
+2: [variable a.x]
+1: [scan a] [index scan a@primary]
+
+# TODO(peter): The index scan and project get mapped to the same group
```

Comment:
I was under the impression that you can't in general tell if two memo expressions are equivalent, and (apart for some simple cases were we can tell) that information would come from transformatons. Eg join commutativity knows that the new expr is supposed to go in the same group.  I may have gotten things wrong. I wonder how Orca does it exactly.

Reply:
Orca keeps track of "constraints" on column values as part of the relational properties. I'm still trying to understand how this work. See `base/CConstraint.h` and the `CConstraintInterval`, `CConstraintConjunction`, `CConstraintDisjunction` and `CContraintNegation` implementations. I'm trying to work out how this overlaps with predicates and why we'd want to keep both formats. I'm also try to figure out how certain scalar expressions get mapped into constraints. For example, `lower(x) > 'foo'`.

In the short term, one thing I've been thinking we might do is to include the "filters" applied at an expression in `relationalProps`. These filters would have to be propagated up from children.

Reply:
That is tricky because the filters may not even be expressible at the level of a given group. Eg a filter below a groupby (SELECT SUM(x) FROM t WHERE y > 1 GROUP BY z). Not sure how y>1 can be expressed at the level of the root group.

Reply:
I think that `y>1` wouldn't be expressed at the root level because the column `y` is not utilized after the selection expression. When an operator is determining its constraints, it would include constraints from its inputs only as long as the constraints were in columns that are part of its outputs.

Reply:
But if there are no constraints, then wouldn't the query that does not have the "y > 1" filter be considered equivalent to the query that does have it (i.e. because the relational props would be the same)?

Reply:
Yes, but is that a problem? Seems like it could be, though I'm having trouble coming up with an example why.

Reply:
Say the `t` table has this data:
```
x     y     z
-------------
100   0     5
100   2     5
```

The two queries (with and without y > 1) would return different results (100 vs. 200). So they're not logically equivalent, even though the output schema is the same.

Reply:
Yeah, I realize the queries are different, but I was failing to see if that was a problem in practice. The answer is a definite "yes" as we can join the two different queries together in a single query using either `UNION` or `JOIN`. This implies that we'd need to propagate the `y > 1` constraint all the way to the root of the expression. Or, we need to add some sort of uniquifier to the relational properties when a constraint is not propagated.